### PR TITLE
Selective pruning using hashmask

### PR DESCRIPTION
**Pull request by Greg-Griffith**
This PR adds random block pruning. 
This PR adds a new configurable param which when enabled, will set the node to prune based on a randomly generated 64bit number (henceforth hashmask). A second configurable param called threshold (integer between 1 and 100, 100 by default) can be set to lower threshold by which blocks get prune. 
formula is as follows:
given 
normalized_threshold = threshold * MAX_UINT64/100
then
if (blockhash_64LSB XOR hashmask) < normalized_threshold
keep block
else 
prune block

__TODO__
- if both prune and prunewithmask are set, return an error with an explaination
- if previously had prune or prune with mask, return an error when switching to the other

### CODE REVIEW

**Comment by sickpig** [comment]
@Greg-Griffith need rebase. w/o it travis won't be executed.

**Comment by Greg-Griffith** [comment]
~this PR is on hold until a decision about XVersion is made since this feature is completely useless without it (or the addition of a different new message type).~

**Comment by sickpig** [comment]
@Greg-Griffith restarted all timed out jobs. You still need to fix a formatting problem thou.

**Comment by Greg-Griffith** [comment]
rebased

**Comment by ptschip** [comment]
Since you resolved your network issue maybe you can drop those last two python test commits...IMO it;s valuable to keep the tests in their original form if we can, for one because they will run faster and two, because they can help to trigger real issues that need debugging.

**Comment by Greg-Griffith** [comment]
> Since you resolved your network issue maybe you can drop those last two python test commits...IMO it;s valuable to keep the tests in their original form if we can, for one because they will run faster and two, because they can help to trigger real issues that need debugging.

I dropped those commits from the PR. 

My counter argument for this would be it would solve some annoyances with travis instability, but you make a compelling point about triggering real issues that need debugging which I agree is more important.

File: qa/rpc-tests/randompruning.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        uint64_t_max = (2**64) - 1
+        threshold_percent = uint64_t_max / 100
+        blockchaininfo = self.nodes[1].getblockchaininfo()
+        fullhashMask = blockchaininfo["pruneHashMask"]
+        threshold = blockchaininfo["hashMaskThreshold"]
+        hashMask64 = GetLow64(fullhashMask)
+        normalized_threshold = threshold * threshold_percent
+
+        kept = []
+        for block in blocks_mined:
```
Comment (ptschip) [inline qa/rpc-tests/randompruning.py:74]: Could maybe add some comments here as to what this test is doing and what the expected results are.

File: src/allowed_args.cpp [hunk 2]
```diff
@@ -310,10 +310,15 @@ static void addGeneralOptions(AllowedArgs &allowedArgs, HelpMessageMode mode)
                     DEFAULT_PERSIST_MEMPOOL))
         .addArg("prune=<n>", requiredInt,
             strprintf(_("Reduce storage requirements by pruning (deleting) old blocks. This mode is incompatible with "
-                        "-txindex and -rescan. "
+                        "-txindex and -rescan. Incompatible with -prunewithmask"
                         "Warning: Reverting this setting requires re-downloading the entire blockchain. "
```
Comment (ptschip) [inline src/allowed_args.cpp:316]: Nit:  Should it maybe better to say "This mode is incompatible with -txindex, -rescan and -prunewithmask"  .

File: src/allowed_args.cpp [hunk 3]
```diff
@@ -310,10 +310,15 @@ static void addGeneralOptions(AllowedArgs &allowedArgs, HelpMessageMode mode)
...
                         "(default: 0 = disable pruning blocks, >%u = target size in MiB to use for block files)"),
                     MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024))
+        .addArg("prunewithmask={true,false,0,1}", optionalBool,
+            strprintf(_("Reduce storage requirements by pruning (deleting) old blocks. This mode is incompatible with "
+                        "-txindex and -rescan. Incompatible with -prune "
+                        "Warning: Reverting this setting requires re-downloading the entire blockchain. "
```
Comment (ptschip) [inline src/allowed_args.cpp:322]: Nit, same this here "This mode is incompatible with -txindex, -rescand and -prune" ?

File: qa/rpc-tests/randompruning.py (hunk 1 above)
Comment (Greg-Griffith) [reply qa/rpc-tests/randompruning.py:74]: added.

File: src/allowed_args.cpp (hunk 2 above)
Comment (Greg-Griffith) [reply src/allowed_args.cpp:316]: fixed

**Comment by ptschip** [comment]
looks like a small formatting issue.  I'll try to look at this in the next few days.

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -1122,15 +1122,20 @@ UniValue getblockchaininfo(const UniValue &params, bool fHelp)
 
     if (fPruneMode)
     {
-        CBlockIndex *block = chainActive.Tip();
+        if (!fPruneWithMask)
         {
-            READLOCK(cs_mapBlockIndex);
+            CBlockIndex *block = chainActive.Tip();
             while (block && block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA))
```
Comment (ptschip) [inline src/rpc/blockchain.cpp:1179]: why take out the READLOCK()? We need that for checking nStatus.

File: src/uint256.cpp [hunk 5]
```diff
@@ -76,6 +77,18 @@ std::string base_blob<BITS>::ToString() const
     return (GetHex());
 }
 
+template <unsigned int BITS>
+std::string base_blob<BITS>::ToBitString()
+{
```
Comment (ptschip) [inline src/uint256.cpp]: interesting but should maybe go in a separate PR

**Comment by ptschip** [comment]
I completed my first pass, will probably need to go over it again.  I'll do some testing first though.

File: qa/rpc-tests/randompruning.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            assert_equal(len(block), 64)
+            low64block = GetLow64(block)
+            # if the 64LSB of the blockhash is equal to or above our threshold we should have pruned it
+            # assert this is the case
+            if ((int(low64block, 16) ^ int(hashMask64, 16)) >= normalized_threshold):
+                assert_raises_rpc_error(-32603, "Block not available (pruned data)", self.nodes[1].getblock, block)
+            else:
+                kept.append(self.nodes[1].getblock(block)['height'])
+
```
Comment (ptschip) [inline qa/rpc-tests/randompruning.py:84]: shouldn't we also check that we kept the blocks we expected to?

File: qa/rpc-tests/randompruning.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            # dont add blocks above the min blocks to keep line
+            if (i < 80):
+                blocks_mined.append(blockhash)
+        self.sync_blocks()
+
+        #get the max possible value of our hashmask
+        uint64_t_max = (2**64) - 1
+        # calculate 1% which
+        threshold_percent = uint64_t_max / 100
+        blockchaininfo = self.nodes[1].getblockchaininfo()
```
Comment (ptschip) [inline qa/rpc-tests/randompruning.py:60]: maybe add a test using prune.hashMaskThreshold=<something> and verify that it works.

Comment (ptschip) [reply qa/rpc-tests/randompruning.py:60]: Also add a test where you change prune.hashMaskThreshold to a lower value and then check whether it's actually pruning the blocks correctly.  I seem to think from testing that this actually isn't working but I'm not entirely sure.

File: src/rpc/blockchain.cpp (hunk 4 above)
Comment (Greg-Griffith) [reply src/rpc/blockchain.cpp:1179]: good catch, i will add that back in.

File: qa/rpc-tests/randompruning.py (hunk 6 above)
Comment (Greg-Griffith) [reply qa/rpc-tests/randompruning.py:84]: doesnt this already do that? if we dont raise the assert where we think we should, then it implies that are keeping everything we expect to keep?

File: qa/rpc-tests/randompruning.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            else:
+                kept.append(self.nodes[1].getblock(block)['height'])
+
+        print("checking if we can raise the threshold higher than the current one... (should not be able to)")
+        # we should get an error raising the threshold higher than 10% since that was its last value
+        assert_raises_rpc_error(-1, "current hashMaskThreshold (10) is smaller than your proposed new threshold (70)",
+        self.nodes[1].set, "prune.hashMaskThreshold=70")
+        # our threshold should still be 10
+        print("checking to make sure nothing has changed...")
+        assert_equal({'prune.hashMaskThreshold': 10}, self.nodes[1].get("prune.hashMaskThreshold"))
```
Comment (gandrewstone) [inline qa/rpc-tests/randompruning.py:91]: Having only 100 divisions seems too few, especially to add a feature where we automatically reduce the threshold periodically to keep the blockchain size < a configured size in GB.  This feature is what people really want since they can set it to whatever is appropriate based on their disk size.

We could expose the raw threshold number, we could accept floating point numbers in the CLI, or we could used fixed point -- that is, the threshold is a percent * 1e6  (i.e. 1% -> prune.hashMaskThreshold=1000000) 

EDIT: I just realized that with fine gradations, the constant GB feature could be implemented as a script...

File: src/allowed_args.cpp [hunk 9]
```diff
@@ -311,10 +312,19 @@ static void addGeneralOptions(AllowedArgs &allowedArgs, HelpMessageMode mode)
...
         .addArg("prune=<n>", requiredInt,
             strprintf(_("Reduce storage requirements by pruning (deleting) old blocks. This mode is incompatible with "
-                        "-txindex and -rescan. "
+                        "-txindex, -rescan, and -prunewithmask"
                         "Warning: Reverting this setting requires re-downloading the entire blockchain. "
                         "(default: 0 = disable pruning blocks, >%u = target size in MiB to use for block files)"),
                     MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024))
+        .addArg("prunewithmask={true,false,0,1}", optionalBool,
+            strprintf(_("Reduce storage requirements by pruning (deleting) old blocks. This mode is incompatible with "
+                        "-txindex, -rescan, and -prune "
```
Comment (gandrewstone) [inline src/allowed_args.cpp:321]: What happens if these flags are both used?  We should make sure no crash.

File: src/allowed_args.cpp [hunk 10]
```diff
@@ -311,10 +312,19 @@ static void addGeneralOptions(AllowedArgs &allowedArgs, HelpMessageMode mode)
...
+                        "Warning: Reverting this setting requires re-downloading the entire blockchain. "
+                        "(default: 0 = disable pruning blocks, 1 = prune blocks using a randomly generated mask)")))
+        .addArg("prunethreshold=<n>", requiredInt,
+            strprintf(_("A normalized threshold to determine how"
+                        "many blocks to keep when prunning with hashmask (default: %u)"),
+                    DEFAULT_THRESHOLD_PERCENT))
```
Comment (gandrewstone) [inline src/allowed_args.cpp]: If you have this as a tweak, you don't need to .addArg.  All tweaks are automatically added as args.

File: src/arith_uint256.cpp [hunk 11]
```diff
@@ -179,6 +180,17 @@ std::string base_uint<BITS>::ToString() const
     return (GetHex());
 }
 
+template <unsigned int BITS>
+std::string base_uint<BITS>::ToBitString() const
```
Comment (gandrewstone) [inline src/arith_uint256.cpp]: Add a unit test for this API with known test vectors

File: src/arith_uint256.cpp [hunk 12]
```diff
@@ -179,6 +180,17 @@ std::string base_uint<BITS>::ToString() const
...
+{
+    std::string bits = "";
+    for (int i = 0; i < WIDTH; i++)
+    {
+        bits = bits + std::bitset<WIDTH>(pn[i]).to_string();
```
Comment (gandrewstone) [inline src/arith_uint256.cpp]: how about bits.append(std::bitset<WIDTH>(pn[i]).to_string()) to make this a little more efficient?  I wish we could hand bitset a bigger array and avoid this whole for loop.

File: src/arith_uint256.h [hunk 13]
```diff
@@ -240,6 +240,7 @@ class base_uint
     void SetHex(const char *psz);
     void SetHex(const std::string &str);
     std::string ToString() const;
+    std::string ToBitString() const;
```
Comment (gandrewstone) [inline src/arith_uint256.h]: add doc comment

File: src/blockstorage/blockleveldb.h [hunk 14]
```diff
@@ -123,6 +128,7 @@ class CBlockLevelDB : public CDatabaseAbstract
         pwrapperblock->Sync();
         pwrapperundo->Sync();
     }
+    bool EraseBlock(const std::string &key);
```
Comment (gandrewstone) [inline src/blockstorage/blockleveldb.h:131]: add doc comments to all new apis

File: src/blockstorage/blockstorage.cpp [hunk 15]
```diff
@@ -796,15 +747,7 @@ bool FindBlockPos(CValidationState &state,
     {
         pos.nFile = 1;
         pos.nPos = nAddSize;
-        if (CheckDiskSpace(nAddSize))
-        {
-            nDBUsedSpace += nAddSize;
-            if (fPruneMode && nDBUsedSpace >= nPruneTarget)
```
Comment (gandrewstone) [inline src/blockstorage/blockstorage.cpp:802]: are you removing the old pruning method?  Its seems like you are not because the flag is still in addargs but a lot of that code is being removed...

File: src/blockstorage/prune.cpp [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+    std::mt19937_64 rand(seed); // Standard mersenne_twister_engine seeded with rd()
+    uint64_t hashMask64(rand());
+    pruneHashMask = 0;
+    pruneHashMask = pruneHashMask | hashMask64;
+    pblocktree->WriteHashMask(pruneHashMask);
+    pblocktree->WriteFlag("hashmaskexists", true);
+}
+
+/** Get important pruning bits from a block hash and compare their value to our pruning threshold*/
+bool hashMaskCompare(uint256 _blockHash)
```
Comment (gandrewstone) [inline src/blockstorage/prune.cpp]: Let's name this better, esp since its not static.  We want to communicate that its about pruning and what is the sense of the returned boolean.  How about:

isBlockPrunable()

Note that if you use this name, it would also swap the return -- that is true would mean prune, and false means keep.

File: src/blockstorage/prune.cpp [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+        }
+        if (pindexOldest->nFile == 0)
+        {
+            // already pruned block
+            pindexOldest = chainActive.Next(pindexOldest);
+            continue;
+        }
+        if (!fPruneWithMask && nDBUsedSpace < nPruneTarget)
+        {
+            break;
```
Comment (gandrewstone) [inline src/blockstorage/prune.cpp:389]: this is implements the old style pruning?

File: src/blockstorage/prune.cpp [hunk 18]
```diff
@@ -0,0 +1,418 @@
...
+        if (pindexOldest->nHeight >= (int)nLastBlockWeCanPrune)
+        {
+            break;
+        }
+        if (fPruneWithMask == true && hashMaskCompare(pindexOldest->GetBlockHash()))
+        {
+            pindexOldest = chainActive.Next(pindexOldest);
+            continue;
+        }
+        unsigned int blockSize = pindexOldest->nDataPos;
```
Comment (gandrewstone) [inline src/blockstorage/prune.cpp]: Add comment here:
// We've chosen to prune this block

File: src/blockstorage/prune.cpp [hunk 19]
```diff
@@ -0,0 +1,418 @@
...
+        if (pindexOldest->nFile == 0)
+        {
+            // already pruned block
+            pindexOldest = chainActive.Next(pindexOldest);
+            continue;
+        }
+        if (!fPruneWithMask && nDBUsedSpace < nPruneTarget)
+        {
+            break;
+        }
```
Comment (gandrewstone) [inline src/blockstorage/prune.cpp:390]: with information like nDBUsedSpace already existing, I feel that it would be quite easy to allow the hashMaskThreshold tweak to take either a % or a disk size.  The hard part would be parsing it but perhaps we could require "%" or "G" to distinguish.  

Now that we have the desired size, if nDBUsedSpace is < desired, we do nothing.  (perhaps in a later PR we have an RPC call that will fill)
If nDBUsedSpace is > desired-fudgefactor (the fudge factor is so that we don't bump over the limit next block), then we figure out how much to decrease hashMaskThreshold by.  First figure out how much we want nDBUsedSpace to fall by:  (desired-fudgefactor)/nDBUsedSpace.  This is a fraction from 0-1.  Now we eliminate (on average) that many blocks:

hashMaskThreshold = hashMaskThreshold*((desired-fudgefactor)/nDBUsedSpace).

That should keep automatically reducing the hashMaskThreshold to keep nDBUsedSpace < desired.

**Comment by gandrewstone** [comment]
We need a "pruning.md" file in docs that basically describes the approach.  

And most importantly explains how to set it up from an end-user perspective.  In other words, what options to use together in the bitcoin.conf file, and what can be modified via tweaks.

Please include the different pruning options (the old style vs the new style).

File: src/allowed_args.cpp (hunk 10 above)
Comment (Greg-Griffith) [reply src/allowed_args.cpp]: removed this entry from args list.

File: src/arith_uint256.cpp (hunk 11 above)
Comment (Greg-Griffith) [reply src/arith_uint256.cpp]: removed from PR as it is unused. will put it in its own PR as ptschip suggested.

File: src/arith_uint256.cpp (hunk 12 above)
Comment (Greg-Griffith) [reply src/arith_uint256.cpp]: see above

File: src/arith_uint256.h (hunk 13 above)
Comment (Greg-Griffith) [reply src/arith_uint256.h]: see above

File: src/blockstorage/prune.cpp (hunk 16 above)
Comment (Greg-Griffith) [reply src/blockstorage/prune.cpp]: it isnt declared as a static function, but it is only used in this one file. I will change the name and remove it from the header.

File: src/blockstorage/blockstorage.cpp (hunk 15 above)
Comment (Greg-Griffith) [reply src/blockstorage/blockstorage.cpp:802]: i did not remove the old pruning method. the code was moved/refactored into prune.cpp to make managing the pruning code easier overall mostly by reducing the amount of public api for pruning code. it is down to only a handful of functions.

File: src/blockstorage/prune.cpp (hunk 17 above)
Comment (Greg-Griffith) [reply src/blockstorage/prune.cpp:389]: in a nutshell, yes. space pruning hasnt been changed, just refactored.

File: src/blockstorage/prune.cpp (hunk 18 above)
Comment (Greg-Griffith) [reply src/blockstorage/prune.cpp]: added

File: src/allowed_args.cpp (hunk 9 above)
Comment (Greg-Griffith) [reply src/allowed_args.cpp:321]: i have reworked SetupPruning to address this concern. All pruning mode checks are now done in this function instead of during various different parts of init

File: src/blockstorage/blockleveldb.h (hunk 14 above)
Comment (Greg-Griffith) [reply src/blockstorage/blockleveldb.h:131]: these apis are documented in the abstract class where i defined the virtual functions

**Comment by gandrewstone** [comment]
After sync, I configured the threshold for 25%.  The CLI refused to work for hours with the message "pruning blockstore...".  Now I'm at 120% CPU (constantly) and the CLI is quite slow.  But neither of these should be the case since pruning just reduces what's available.  So it should be implementable as a background thread that grabs cs_main when it needs to prune a block, deletes the block, then gives cs_main back, and then does a short delay.  

We should not let pruning interrupt normal processing, since all its doing is reducing the disk space.
